File: software/mobile/ios/SVCMobile/Tests/MockBLETransportTests.swift
import XCTest
@testable import SVCMobile

final class MockBLETransportTests: XCTestCase {
    func testMockScanConnectAndWrite() {
        let transport = MockBLETransport()
        var devices: [DiscoveredDevice] = []
        var state = ConnectionState.disconnected
        transport.onDevicesChanged = { devices = $0 }
        transport.onConnectionChanged = { state = $0 }

        transport.startScan()
        XCTAssertEqual(devices, [transport.device])
        XCTAssertEqual(state, .scanning)

        transport.connect(to: transport.device.id)
        XCTAssertEqual(state, .connected)

        let characteristic = UUID()
        transport.write(Data([1, 2, 3]), characteristic: characteristic)
        XCTAssertEqual(transport.writes.count, 1)
        XCTAssertEqual(transport.writes[0].0, Data([1, 2, 3]))
        XCTAssertEqual(transport.writes[0].1, characteristic)
    }
}

File: software/mobile/android/core-ble/src/test/kotlin/com/avlyubimov/svc/core/ble/MockBleTransportTests.kt
package com.avlyubimov.svc.core.ble

import com.avlyubimov.svc.core.model.ConnectionState
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals

class MockBleTransportTests {
    @Test
    fun scansConnectsAndRecordsWrites() {
        val transport = MockBleTransport()
        transport.startScan()
        assertEquals(ConnectionState.SCANNING, transport.connectionState.value)
        val device = transport.devices.value.single()
        transport.connect(device.id)
        assertEquals(ConnectionState.CONNECTED, transport.connectionState.value)
        transport.write("9f9e7101-7b7a-4f4f-a001-535643000001", byteArrayOf(1, 2))
        assertEquals(1, transport.writes.size)
        assertContentEquals(byteArrayOf(1, 2), transport.writes.single().second)
    }
}
